react-components/src/components/RuleBasedOutputs/utils.ts

```diff
@@ -0,0 +1,242 @@
+      break;
+    }
+    case 'outside': {
+      const lower = condition.lowerBoundExclusive;
+      const upper = condition.upperBoundExclusive;
+      const value = Number(asset[trigger.type]?.[trigger.key]);
+      expressionResult = value <= lower && upper <= value;
+      break;
+    }
+  }
+
+  return expressionResult;
+};
+
+const traverseExpression = (
+  asset: Asset,
+  expressions: Expression[]
+  // levelExpressionOperator: string
+): boolean[] => {
+  let expressionResult: boolean = false;
+
+  const expressionResults: boolean[] = [];
+
+  expressions.forEach((expression) => {
+    switch (expression.type) {
+      case 'or': {
+        const operatorResult = traverseExpression(
+          asset,
+          expression.expressions
+          // expression.type
+        );
+        expressionResult = operatorResult.find((result) => result) ?? false;
+        break;
+      }
+      case 'and': {
+        const operatorResult = traverseExpression(
+          asset,
+          expression.expressions
+          // expression.type
+        );
+        expressionResult = operatorResult.find((result) => !result) ?? false;
+        break;
+      }
+      case 'not': {
+        const operatorResult = traverseExpression(
+          asset,
+          [expression.expression]
+          // expression.type
+        );
+        expressionResult = !operatorResult[0]; // TODO: make the not operator
+        break;
+      }
+      case 'numericExpression': {
+        expressionResult = checkNumericExpressionStatement(asset, expression);
+        break;
+      }
+      case 'stringExpression': {
+        expressionResult = checkStringExpressionStatement(asset, expression);
+        break;
+      }
+      default: {
+        // statements;
+        break;
+      }
+    }
+    expressionResults.push(expressionResult);
+  });
+
+  return expressionResults;
+};
+
+export const generateRuleBasedOutputs = (
+  model: CogniteCadModel,
+  contextualizedAssetNodes: Asset[],
+  assetMappings: AssetMapping3D[],
+  ruleSet: RuleOutputSet
+): void => {
+  const outputType = 'color'; // for now it only supports colors as the output
+
+  type RuleAndStyleIndex = {
+    styleIndex: TreeIndexNodeCollection;
+    ruleOutputParams: RuleOutput;
+  };
+
+  const ruleWithOutputs = ruleSet?.rulesWithOutputs;
+
+  ruleWithOutputs?.forEach((ruleWithOutput: { rule: any; outputs: any }) => {
+    const { rule, outputs } = ruleWithOutput;
+    // Starting Expression
+    const expression = rule.expression;
+
+    const outputSelected = outputs.find(
+      (output: { type: string }) => output.type === outputType
+    ) as ColorRuleOutput;
+
+    const ruleOutputAndStyleIndex: RuleAndStyleIndex = {
+      styleIndex: new TreeIndexNodeCollection(),
+      ruleOutputParams: outputSelected
+    };
+
+    if (outputSelected === undefined) {
+      console.log('No rule output found for the type requested: ', outputType, outputs);
```

Comment:
We can't usually expect users to see messages in the console log, so this could be removed (unless it's important, perhaps we can throw an error?)

Reply:
I think it is better to just remove it because it is more a internal log when traversing the expression types

Reply:
removed here:
https://github.com/cognitedata/reveal/pull/4210/commits/d748037cc48301eb04ef215bd184c86e4158a622